test/test-thread-pool.cpp

```diff
@@ -0,0 +1,796 @@
+    for (size_t i = 0; i < handlers.size(); i++) {
+      auto handler = handlers.at(i);
+      if (handler->state.load() == STATE::DONE) {
+        counter++;
+      }
+    }
+    // Account for thread pools of 3 threads as tasks are added in pairs
+    ASSERT_NEAR(counter, ThreadPoolExecutor::GetNumberOfThreads(), 1);
+    ASSERT_EQ(tasks_counter.load(), 4);
+  }
+
+  // Nested add task
+  {
+    sync.store(2);
+    tasks_counter.store(0);
+    ThreadPoolExecutor::SetNumberOfThreads(6);
+
+    ThreadPoolExecutor::AddRecursiveCalls(
+        [&](int id, int threads) {
+          HEXL_UNUSED(id);
+          HEXL_UNUSED(threads);
+          tasks_counter.fetch_add(1);
+          ThreadPoolExecutor::AddRecursiveCalls(add_task, add_task);
+        },
+        [&](int id, int threads) {
+          HEXL_UNUSED(id);
+          HEXL_UNUSED(threads);
+          tasks_counter.fetch_add(1);
+          ThreadPoolExecutor::AddRecursiveCalls(add_task, add_task);
+        });
+
+    int counter = 0;
+    auto handlers = ThreadPoolExecutor::GetThreadHandlers();
+    for (size_t i = 0; i < handlers.size(); i++) {
+      auto handler = handlers.at(i);
+      if (handler->state.load() == STATE::DONE) {
+        counter++;
+      }
+    }
+    // Account for thread pools of 3 or 5 threads as tasks are added in pairs
+    ASSERT_NEAR(counter, ThreadPoolExecutor::GetNumberOfThreads(), 1);
+    ASSERT_EQ(tasks_counter.load(), 6);
+  }
+
+  // Add task & stop threads in parallel
+  {
+    sync.store(2);
+    tasks_counter.store(0);
+    ThreadPoolExecutor::SetNumberOfThreads(0);
+    std::thread thread_object1([&]() {
+      sync.fetch_add(-1);
+      while (sync) {
+      }
+      ThreadPoolExecutor::AddRecursiveCalls(add_task, add_task);
+    });
+    std::thread thread_object2([&sync]() {
+      sync.fetch_add(-1);
+      while (sync) {
+      }
+      ThreadPoolExecutor::SetNumberOfThreads(0);
+    });
+
+    thread_object1.join();
+    thread_object2.join();
+
+    uint64_t pool_size = ThreadPoolExecutor::GetNumberOfThreads();
+    ASSERT_TRUE(pool_size == 0 || pool_size == HEXL_NUM_THREADS ||
+                pool_size == std::thread::hardware_concurrency());
+    ASSERT_EQ(tasks_counter.load(), 2);
+  }
+
+  // Parallel add jobs
+  {
+    sync.store(2);
+    iterations.store(0);
+    int N_size = 100;
+    ThreadPoolExecutor::SetNumberOfThreads(2);
+    std::thread thread_object1([&]() {
+      sync.fetch_add(-1);
+      while (sync) {
+      }
+      ThreadPoolExecutor::AddParallelJobs(add_iterations);
+    });
+    std::thread thread_object2([&]() {
+      sync.fetch_add(-1);
+      while (sync) {
+      }
+      ThreadPoolExecutor::AddParallelJobs(add_iterations);
+    });
+
+    thread_object1.join();
+    thread_object2.join();
+
+    int counter = 0;
+    auto handlers = ThreadPoolExecutor::GetThreadHandlers();
+    for (size_t i = 0; i < handlers.size(); i++) {
+      auto handler = handlers.at(i);
```

Comment:
```suggestion
      auto handler = handlers[i];
```
[] does no bounds check but that is ok because your loop is bounded.